network/src/ncp/handler.rs

```diff
@@ -21,110 +21,238 @@
+    /// Configuration for the Ncp protocol.
+    config: NcpConfig,
+
+    /// If true, we are trying to shut down the existing NCP substream and should refuse any
+    /// incoming connection.
+    shutting_down: bool,
+
+    /// The active substreams.
+    // TODO: add a limit to the number of allowed substreams
+    substreams: Vec<SubstreamState<TSubstream>>,
+
+    /// Queue of values that we want to send to the remote.
+    send_queue: SmallVec<[NcpMessage; 16]>,
+}
+
+/// State of an active substream, opened either by us or by the remote.
+enum SubstreamState<TSubstream>
+where
+    TSubstream: AsyncRead + AsyncWrite,
+{
+    /// Waiting for a message from the remote.
+    WaitingInput(Framed<TSubstream, NcpCodec>),
+    /// Waiting to send a message to the remote.
+    PendingSend(Framed<TSubstream, NcpCodec>, NcpMessage),
+    /// Waiting to flush the substream so that the data arrives to the remote.
+    PendingFlush(Framed<TSubstream, NcpCodec>),
+    /// The substream is being closed.
+    Closing(Framed<TSubstream, NcpCodec>),
+}
+
+impl<TSubstream> SubstreamState<TSubstream>
+where
+    TSubstream: AsyncRead + AsyncWrite,
+{
+    /// Consumes this state and produces the substream.
+    fn into_substream(self) -> Framed<TSubstream, NcpCodec> {
+        match self {
+            SubstreamState::WaitingInput(substream) => substream,
+            SubstreamState::PendingSend(substream, _) => substream,
+            SubstreamState::PendingFlush(substream) => substream,
+            SubstreamState::Closing(substream) => substream,
+        }
+    }
+}
+
+impl<TSubstream> NcpHandler<TSubstream>
+where
+    TSubstream: AsyncRead + AsyncWrite,
+{
+    /// Builds a new `NcpHandler`.
+    pub fn new() -> Self {
+        NcpHandler {
+            config: NcpConfig::new(),
+            shutting_down: false,
+            substreams: Vec::new(),
+            send_queue: SmallVec::new(),
+        }
+    }
+}
+
+impl<TSubstream> ProtocolsHandler for NcpHandler<TSubstream>
+where
+    TSubstream: AsyncRead + AsyncWrite,
+{
+    type InEvent = NcpMessage;
+    type OutEvent = NcpMessage;
+    type Error = io::Error;
+    type Substream = TSubstream;
+    type InboundProtocol = NcpConfig;
+    type OutboundProtocol = NcpConfig;
+    type OutboundOpenInfo = NcpMessage;
+
+    #[inline]
+    fn listen_protocol(&self) -> Self::InboundProtocol {
+        self.config.clone()
+    }
+
+    fn inject_fully_negotiated_inbound(
+        &mut self,
+        protocol: <Self::InboundProtocol as InboundUpgrade<TSubstream>>::Output,
+    ) {
+        if self.shutting_down {
+            return ();
+        }
+        self.substreams.push(SubstreamState::WaitingInput(protocol))
+    }
+
+    fn inject_fully_negotiated_outbound(
+        &mut self,
+        protocol: <Self::OutboundProtocol as OutboundUpgrade<TSubstream>>::Output,
+        message: Self::OutboundOpenInfo,
+    ) {
+        if self.shutting_down {
+            return ();
```

Comment:
You probably can omit `()` when returning unit type.

```suggestion
            return;
```

Reply:
This is a copy-paste from floodsub code. I'm fine with both variants.